grr/server/grr_response_server/databases/spanner_flows.py

```diff
@@ -0,0 +1,2510 @@
+        request_ids.append(request_id)
+
+      query = (
+        "UPDATE MessageHandlerRequests "
+        "SET LeasedUntil=@leased_until, LeasedBy=@leased_by "
+        "WHERE RequestId IN UNNEST(@request_ids)"
+      )
+      params = {
+        "request_ids": request_ids,
+        "leased_by": leased_by,
+        "leased_until": delivery_time.AsDatetime()
+      }
+      param_type = {
+        "request_ids": param_types.Array(param_types.STRING),
+        "leased_by": param_types.STRING,
+        "leased_until": param_types.TIMESTAMP
+      }
+      txn.execute_update(query, params, param_type)
+
+      return res
+
+    return self.db.Transact(Txn, txn_tag="_LeaseMessageHandlerRequests")
+
+  @db_utils.CallLogged
+  @db_utils.CallAccounted
+  def WriteMessageHandlerRequests(
+      self, requests: Iterable[objects_pb2.MessageHandlerRequest]
+  ) -> None:
+    """Writes a list of message handler requests to the queue."""
+    def Mutation(mut) -> None:
+      creation_timestamp = spanner_lib.COMMIT_TIMESTAMP
+      rows = []
+      columns = ["RequestId", "HandlerName", "CreationTime", "Payload"]
+      for request in requests:
+        rows.append([
+          str(request.request_id),
+          request.handler_name,
+          creation_timestamp,
+          request,
+        ])
+      mut.insert(table="MessageHandlerRequests", columns=columns, values=rows)
+
+    self.db.Mutate(Mutation, txn_tag="WriteMessageHandlerRequests")
+
+  @db_utils.CallLogged
+  @db_utils.CallAccounted
+  def ReadMessageHandlerRequests(
+      self,
+  ) -> Sequence[objects_pb2.MessageHandlerRequest]:
+    """Reads all message handler requests from the queue."""
+    query = """
+    SELECT t.Payload, t.CreationTime, t.LeasedBy, t.LeasedUntil FROM MessageHandlerRequests AS t
+    """
+    results = []
+    for payload, creation_time, leased_by, leased_until in self.db.ParamQuery(query, {}):
+      req = objects_pb2.MessageHandlerRequest()
+      req.ParseFromString(payload)
+      req.timestamp = rdfvalue.RDFDatetime.FromDatetime(
+          creation_time
+      ).AsMicrosecondsSinceEpoch()
+      if leased_by is not None:
+        req.leased_by = leased_by
+      if leased_until is not None:
+        req.leased_until = rdfvalue.RDFDatetime.FromDatetime(
+          creation_time
+      ).AsMicrosecondsSinceEpoch()
+      results.append(req)
+
+    return results
+
+  @db_utils.CallLogged
+  @db_utils.CallAccounted
+  def DeleteMessageHandlerRequests(
+      self,
+      requests: Iterable[objects_pb2.MessageHandlerRequest],
+  ) -> None:
+    """Deletes a list of message handler requests from the database."""
+
+    query = "DELETE FROM MessageHandlerRequests WHERE RequestId IN UNNEST(@request_ids)"
+    request_ids = []
+    for r in requests:
+        request_ids.append(str(r.request_id))
+    params={"request_ids": request_ids}
+    param_type={"request_ids": param_types.Array(param_types.STRING)}
+
+    self.db.ParamExecute(query, params, param_type)
```

Comment:
```suggestion

    self.db.ParamExecute(query, params, txn_tag="DeleteMessageHandlerRequests")
```

`ParamExecute` will derive the type. You want a transaction_tag here instead.